src/testnet/keychain.rs

```diff
@@ -0,0 +1,147 @@
+use std::collections::HashMap;
+
+use chainstate::stacks::{StacksTransactionSigner, TransactionAuth, StacksPublicKey, StacksPrivateKey, StacksAddress};
+use address::AddressHashMode;
+use burnchains::{BurnchainSigner, BurnchainHeaderHash, PrivateKey};
+use util::vrf::{VRF, VRFProof, VRFPublicKey, VRFPrivateKey};
+use util::hash::{Sha256Sum};
+
+pub struct Keychain {
+    secret_keys: Vec<StacksPrivateKey>, 
+    threshold: u16,
+    hash_mode: AddressHashMode,
+    pub hashed_secret_state: Sha256Sum,
+    microblocks_secret_keys: Vec<StacksPrivateKey>,
+    vrf_secret_keys: Vec<VRFPrivateKey>,
+    vrf_map: HashMap<VRFPublicKey, VRFPrivateKey>,
+}
+
+impl Keychain {
+
+    pub fn new(secret_keys: Vec<StacksPrivateKey>, threshold: u16, hash_mode: AddressHashMode) -> Keychain {
+        // Compute hashed secret state
+        let hashed_secret_state = {
+            let mut buf : Vec<u8> = secret_keys.iter()
+                .flat_map(|sk| sk.to_bytes())
+                .collect();
+            buf.extend_from_slice(&[(threshold >> 8) as u8, (threshold & 0xff) as u8, hash_mode as u8]);
+            Sha256Sum::from_data(&buf[..])
+        };
+
+        Self {
+            hash_mode,
+            hashed_secret_state,
+            microblocks_secret_keys: vec![],
+            secret_keys,
+            threshold,
+            vrf_secret_keys: vec![],
+            vrf_map: HashMap::new(),
+        }
+    }
+
+    pub fn default() -> Keychain {
+        let seed_hashed = Sha256Sum::from_data(&[0u8; 32]);
+        let secret_key = StacksPrivateKey::from_slice(seed_hashed.as_bytes()).unwrap();
+        let threshold = 1;
+        let hash_mode = AddressHashMode::SerializeP2PKH;
+
+        Keychain::new(vec![secret_key], 1, hash_mode)
+    }
+
+    pub fn rotate_vrf_keypair(&mut self) -> VRFPublicKey {
+        let seed = match self.vrf_secret_keys.last() {
+            // First key is the hash of the secret state
+            None => self.hashed_secret_state,
+            // Next key is the hash of the last
+            Some(last_vrf) => Sha256Sum::from_data(last_vrf.as_bytes()),  
+        };
+        let sk = VRFPrivateKey::from_bytes(seed.as_bytes()).unwrap();
```

Comment:
There's a chance this can fail if the bytes in the seed don't correspond to a valid Ed25519 private key.  What you can do here to avoid this is to continuously generate seeds through re-hashing until one works.  This ought to keep the key generation deterministic.

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1202/commits/c6072f82031915da18b451253588d9abb2db4993